File: dci-adk-sample/src/main/java/com/domainmodeling/dci/adk/sample/java/moneytransfer/AccountTransferer.java
package com.domainmodeling.dci.adk.sample.java.moneytransfer;

import com.domainmodeling.dci.adk.core.role.AbstractDCIRole;

import java.math.BigDecimal;

public class AccountTransferer extends AbstractDCIRole<ImmutableAccount> {

    public AccountTransferer(ImmutableAccount data) {
        super(data);
    }

    public AccountLedger income(BigDecimal amount) {
        return ImmutableAccountLedger.builder()
                .account(ImmutableAccount.copyOf(data())
                        .withBalance(data().balance().add(amount)))
                .income(amount)
                .originalBalance(data().balance())
                .build();
    }

    public AccountLedger outcome(BigDecimal amount) {
        return ImmutableAccountLedger.builder()
                .account(ImmutableAccount.copyOf(data())
                        .withBalance(data().balance().subtract(amount)))
                .outcome(amount)
                .originalBalance(data().balance())
                .build();
    }

}

File: dci-adk-sample/src/main/kotlin/com/domainmodeling/dci/adk/sample/kotlin/moneytransfer/AccountTransferer.kt
package com.domainmodeling.dci.adk.sample.kotlin.moneytransfer

import java.math.BigDecimal

interface AccountTransferer {
    fun Account.income(amount: BigDecimal): AccountLedger {
        return AccountLedger(
                this.copy(
                        this.balance.plus(
                                amount
                        )
                ), this.balance, amount, BigDecimal.ZERO
        );
    }

    fun Account.outcome(amount: BigDecimal): AccountLedger {
        return AccountLedger(
                this.copy(
                        this.balance.subtract(
                                amount
                        )
                ), this.balance, BigDecimal.ZERO, amount
        );
    }
}